[Merged by Bors] - feat(ring_theory/roots_of_unity): Roots of unity as union of primitive roots

riccardobrasca opened the pull request:
I have added some lemmas about roots of unity, especially `root_of_unity_eq_uniun_prim` that says that, if there is a primitive `n`-th root of unity in `R`, then the set of `n`-th roots of unity is equal to the union of `primitive_roots i R` for `i | n`.

I will use this lemma in to develop the theory of cyclotomic polynomials.

---

All comments are welcome!

riccardobrasca committed:
chore(src/data/pnat/basic.lean ): add an easy lemma about divisors

riccardobrasca committed:
chore(src/number_theory/divisors.lean): add an equivalent definition of nat.divisors n that is used in nat.sum_totient

riccardobrasca committed:
feat(src/ring_theory/roots_of_unity.lean): add some lemmas

riccardobrasca committed:
Update divisors.lean

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,42 @@ begin
 
+/--The cardinality of the multiset `nth_roots ↑n (1 : R)` is `n`
+if there is a primitive root of unity in `R`. -/
+lemma card_nthroots {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) : (nth_roots ↑n (1 : R)).card = n :=
```
```suggestion
lemma card_nth_roots {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) : (nth_roots ↑n (1 : R)).card = n :=
```

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,42 @@ begin
+
+/--The multiset `nth_roots ↑n (1 : R)` has no repeated elements
+if there is a primitive root of unity in `R`. -/
+lemma nroots_nodup {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) : (nth_roots ↑n (1 : R)).nodup :=
```
```suggestion
lemma nth_roots_nodup {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) : (nth_roots ↑n (1 : R)).nodup :=
```

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
 end
 
+/-The sets `primitive_roots k R` are pairwise disjoint. -/
+lemma prim_roots_disjoint {k l : ℕ} (hk : 0 < k) (hl : 0 < l) (h : k ≠ l) :
```
```suggestion
lemma primitive_roots_disjoint {k l : ℕ} (hk : 0 < k) (hl : 0 < l) (h : k ≠ l) :
```

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
+
+/-If in `R` there is a `n`-th primitive root of unity and `b` divides `n`,
+then there is in `R` a `b`-th primitive root of unity. -/
+lemma prim_of_pow_prim {ζ : R} {n : ℕ+} {a b : ℕ} (h : is_primitive_root ζ n) (hprod : ↑n = a * b)
```
This should go in the `is_primitive_root` namespace: `is_primitive_root.pow` to enable dot notation.
Also, I think it might be better to have `n : nat` instead of `n : pnat`, and then add the condition `0 < a`.

riccardobrasca replied:
It now has `n : nat` instead of `n : pnat`.

It already is in the `is_primitive_root`, isn't it? Or you mean that it's better to change its name?

jcommelin replied:
Right, I meant the name change

riccardobrasca replied:
So you suggest to call it simply `pow` (plus the namespace of course)?

jcommelin replied:
Yup.

riccardobrasca replied:
Done.

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
+    exact (mul_dvd_mul_iff_left hzeroa).1 hdiv }
+end
+
+/-`nth_roots n` as a finset is equal to the uniun of `primitive_roots i R` for `i ∣ n`
```
```suggestion
/-`nth_roots n` as a finset is equal to the union of `primitive_roots i R` for `i ∣ n`
```

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
+/-`nth_roots n` as a finset is equal to the uniun of `primitive_roots i R` for `i ∣ n`
+if there is a primitive root of unity in `R`. -/
+lemma root_of_unity_eq_uniun_prim {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
+  (⟨nth_roots n (1 : R) , nroots_nodup h⟩ : finset R) =
```
Can you please explain why you don't use `roots_of_unity` here? That's what the name suggests...

riccardobrasca replied:
This is just because of my ignorance and I do not know how to consider `roots_of_unity n R` as a `finset`. If you can give me an hint on how to do this I will be happy to modify the rest.

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update roots_of_unity.lean

riccardobrasca committed:
Update roots_of_unity.lean

eric-wieser commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
+
+/-`nth_roots n` as a finset is equal to the union of `primitive_roots i R` for `i ∣ n`
+if there is a primitive root of unity in `R`. -/
+lemma root_of_unity_eq_uniun_prim {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
```
```suggestion
lemma root_of_unity_eq_union_prim {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
```

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
+          = x ^ (a * d)  : by rw hd
+      ... = (x ^ a) ^ d  : pow_mul x a d
+      ... = 1 ^ d        : by rw ((mem_primitive_roots hazero).1 ha.2).1
+      ... = 1            : one_pow d },
```
Indent is misleading here,
```suggestion
    { by_contra h,
      simp only [le_zero_iff_eq, not_lt] at h,
      simp only [h, zero_mul, pnat.ne_zero] at hd,
      exact hd },
    calc  x ^ ↑n
        = x ^ (a * d)  : by rw hd
    ... = (x ^ a) ^ d  : pow_mul x a d
    ... = 1 ^ d        : by rw ((mem_primitive_roots hazero).1 ha.2).1
    ... = 1            : one_pow d },
```

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +618,88 @@ begin
+
+/-`nth_roots n` as a finset is equal to the union of `primitive_roots i R` for `i ∣ n`
+if there is a primitive root of unity in `R`. -/
+lemma root_of_unity_eq_union_prim {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
```
```suggestion
lemma nth_roots_one_eq_bind_primitive_roots {ζ : R} {n : ℕ+} (h : is_primitive_root ζ n) :
```
Ok, I see why it's easier/better to use `nth_roots` instead of `roots_of_unity`. So let's reflect that in the name.

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Add definition of `nth_roots_finset`

`nth_roots_finset` is the same as `nth_roots` but viewed as a `finset` (when it does not have duplicate elements).

riccardobrasca committed:
Update roots_of_unity.lean

riccardobrasca committed:
Update roots_of_unity.lean

riccardobrasca committed:
chore(src/ring_theory/roots_of_unity.lean): improve definition of nth_roots_finset

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,46 @@ begin
   exact h.card_roots_of_unity'
 end
 
+/--The cardinality of the multiset `nth_roots ↑n (1 : R)` is `n`
```
```suggestion
/-- The cardinality of the multiset `nth_roots ↑n (1 : R)` is `n`
```

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,46 @@ begin
+    exact hcard }
+end
+
+/--The multiset `nth_roots ↑n (1 : R)` has no repeated elements
```
```suggestion
/-- The multiset `nth_roots ↑n (1 : R)` has no repeated elements
```

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,46 @@ begin
+
+/--The multiset `nth_roots ↑n (1 : R)` as a finset. -/
+def nth_roots_finset (n : ℕ+) (R : Type*) [integral_domain R] : finset R :=
+  multiset.to_finset (nth_roots n (1 : R))
```
```suggestion
/-- The multiset `nth_roots ↑n (1 : R)` as a finset. -/
def nth_roots_finset (n : ℕ+) (R : Type*) [integral_domain R] : finset R :=
multiset.to_finset (nth_roots n (1 : R))
```

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +622,88 @@ begin
     exact ⟨i, ⟨hin, hi.symm⟩, H⟩ }
 end
 
+/-The sets `primitive_roots k R` are pairwise disjoint. -/
```
```suggestion
/-- The sets `primitive_roots k R` are pairwise disjoint. -/
```

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +622,88 @@ begin
+/-If in `R` there is a `n`-th primitive root of unity and `b` divides `n`,
+then there is in `R` a `b`-th primitive root of unity. -/
+lemma prim_of_pow_prim {ζ : R} {n : ℕ} {a b : ℕ} (hn: 0 < n) (h : is_primitive_root ζ n) (hprod : n = a * b)
+  : is_primitive_root (ζ ^ a) b :=
```
```suggestion
/-- If there is a `n`-th primitive root of unity in `R` and `b` divides `n`,
then there is a `b`-th primitive root of unity in `R`. -/
lemma prim_of_pow_prim {ζ : R} {n : ℕ} {a b : ℕ} 
  (hn: 0 < n) (h : is_primitive_root ζ n) (hprod : n = a * b) :
  is_primitive_root (ζ ^ a) b :=
```

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +622,88 @@ begin
+    exact (mul_dvd_mul_iff_left hzeroa).1 hdiv }
+end
+
+/-`nth_roots n` as a finset is equal to the union of `primitive_roots i R` for `i ∣ n`
```
```suggestion
/-- `nth_roots n` as a `finset` is equal to the union of `primitive_roots i R` for `i ∣ n`
```

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

riccardobrasca committed:
Update roots_of_unity.lean

hrmacbeth commented on src/number_theory/divisors.lean:
```diff
@@ -248,4 +248,24 @@ lemma prod_divisors_prime_pow {α : Type*} [comm_monoid α] {k p : ℕ} {f : ℕ
+    { simp only [nat.mem_divisors, and_true, ne.def, pnat.ne_zero, not_false_iff] at hx,
+      exact hx } },
+  {
+    intros x hx,
```
```suggestion
  { intros x hx,
```

hrmacbeth commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,46 @@ begin
+      simp only [multiset.mem_erase_dup, finset.mem_mk] },
+    rw [hrw, ← fintype.card_congr (roots_of_unity_equiv_nth_roots R n), card_roots_of_unity h] at ha,
+    exact nat.lt_asymm ha ha
+  }
```
```suggestion
    exact nat.lt_asymm ha ha }
```

hrmacbeth commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +622,89 @@ begin
+      obtain ⟨d, hd⟩ := hx.2,
+      rw mul_comm at hd,
+      exact card_primitive_roots (pow (pnat.pos n) h hd) (pnat.pos_of_div_pos hx.2) },
+    },
```
```suggestion
      exact card_primitive_roots (pow (pnat.pos n) h hd) (pnat.pos_of_div_pos hx.2) } },
```

hrmacbeth reviewed (commented):
Just whitespace!  Looks great.

riccardobrasca committed:
Update src/number_theory/divisors.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

hrmacbeth commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -559,6 +560,45 @@ begin
+  { exact card_nth_roots n 1 },
+  { rw [not_lt],
+    have hcard : fintype.card {x // x ∈ nth_roots n (1 : R)}
+    ≤ (nth_roots n (1 : R)).attach.card := multiset.card_le_of_le (multiset.erase_dup_le _),
```
```suggestion
      ≤ (nth_roots n (1 : R)).attach.card := multiset.card_le_of_le (multiset.erase_dup_le _),
```

hrmacbeth commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +621,88 @@ begin
+  split,
+  { calc (ζ ^ a) ^ b
+       = ζ ^ n : by rw [← pow_mul', mul_comm, hprod]
+   ... = 1 : by rw ((iff_def ζ n).1 h).1 },
```
```suggestion
        = ζ ^ n : by rw [← pow_mul', mul_comm, hprod]
    ... = 1 : by rw ((iff_def ζ n).1 h).1 },
```

hrmacbeth commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +621,88 @@ begin
+      simp,
+      obtain ⟨d, hd⟩ := hx.2,
+      rw mul_comm at hd,
+      exact card_primitive_roots (pow (pnat.pos n) h hd) (pnat.pos_of_div_pos hx.2) } },
```
if I am mentally parsing correctly, should lines 692-700 inclusive have an extra indent?

hrmacbeth commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +621,88 @@ begin
+    { symmetry; exact nat.sum_totient n },
+    { intros x hx,
+      simp only [finset.mem_filter, finset.mem_range] at hx,
+      simp,
```
nonterminal simp (you can use `squeeze_simp` to replace it)

hrmacbeth reviewed (commented):
A couple more small style points.

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

kbuzzard commented on src/data/pnat/basic.lean:
```diff
@@ -343,6 +343,15 @@ theorem dvd_antisymm {m n : ℕ+} : m ∣ n → n ∣ m → m = n :=
+  rw [not_not] at hzero,
+  rw hzero at h,
+  exact pnat.ne_zero n (eq_zero_of_zero_dvd h)
+end
```
```suggestion
begin
  apply zero_lt_iff_ne_zero.2,
  intro hzero,
  rw hzero at h,
  exact pnat.ne_zero n (eq_zero_of_zero_dvd h)
end
```

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

riccardobrasca committed:
Update src/data/pnat/basic.lean

Co-authored-by: Kevin Buzzard <k.buzzard@imperial.ac.uk>

riccardobrasca committed:
Update roots_of_unity.lean

riccardobrasca committed:
Update roots_of_unity.lean

jcommelin commented on src/number_theory/divisors.lean:
```diff
@@ -248,4 +248,23 @@ lemma prod_divisors_prime_pow {α : Type*} [comm_monoid α] {k p : ℕ} {f : ℕ
   ∏ x in (p ^ k).divisors, f x = ∏ x in range (k + 1), f (p ^ x) :=
 @sum_divisors_prime_pow (additive α) _ _ _ _ h
 
+lemma finset_div_eq_filter (n : ℕ+) :
```
I think I would swap the two sides, make this a simp lemma, and call it `filter_dvd_eq_divisors`.
What do you think?

jcommelin replied:
Also, can't you prove this without splitting, by using `finset.ext` and then `mem_divisors`, `mem_filter`, and `mem_range`?
At least, simping with those before splitting seems to deduplicate some of the proof.

riccardobrasca replied:
Oh, the proof is now much better, thank you! I also swapped the two sides and added `@[simp]`.

jcommelin replied:
Could you please also update the name?

riccardobrasca replied:
Oh, of course, sorry.

riccardobrasca committed:
chore(number_theory/divisors.lean): improve finset_div_eq_filter

riccardobrasca committed:
Merge branch 'prim_roots' of github.com:leanprover-community/mathlib into prim_roots

riccardobrasca committed:
Update divisors.lean

riccardobrasca committed:
chore(ring_theory/roots_of_unity.lean): correct proof of nth_roots_one_eq_bind_primitive_roots

riccardobrasca committed:
Merge branch 'prim_roots' of github.com:leanprover-community/mathlib into prim_roots

riccardobrasca committed:
Update divisors.lean

riccardobrasca committed:
Update roots_of_unity.lean

jcommelin commented on src/number_theory/divisors.lean:
```diff
@@ -248,4 +248,15 @@ lemma prod_divisors_prime_pow {α : Type*} [comm_monoid α] {k p : ℕ} {f : ℕ
+begin
+  apply finset.ext,
+  simp only [mem_divisors, mem_filter, mem_range, and_true,
+  and_iff_right_iff_imp, ne.def, pnat.ne_zero, not_false_iff],
```
```suggestion
    and_iff_right_iff_imp, ne.def, pnat.ne_zero, not_false_iff],
```

jcommelin commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +621,88 @@ begin
+    apply le_of_eq,
+    rw finset.card_bind,
+    {
+      rw ← @finset.sum_congr _ _ (finset.filter (λ (x : ℕ), x ∣ n)
```
```suggestion
    { rw ← @finset.sum_congr _ _ (finset.filter (λ (x : ℕ), x ∣ n)
```

riccardobrasca committed:
Update src/number_theory/divisors.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

jcommelin committed:
some golfing

jcommelin committed:
todo done

bryangingechen commented on src/ring_theory/roots_of_unity.lean:
```diff
@@ -581,6 +624,63 @@ begin
+/-- If there is a `n`-th primitive root of unity in `R` and `b` divides `n`,
+then there is a `b`-th primitive root of unity in `R`. -/
+lemma pow {ζ : R} {n : ℕ} {a b : ℕ}
+  (hn: 0 < n) (h : is_primitive_root ζ n) (hprod : n = a * b) :
```
```suggestion
  (hn : 0 < n) (h : is_primitive_root ζ n) (hprod : n = a * b) :
```

riccardobrasca committed:
Update src/ring_theory/roots_of_unity.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>
